File: app/src/main/java/com/fieldservice/app/presentation/home/HomeViewModel.java
package com.fieldservice.app.presentation.home;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.Transformations;
import androidx.lifecycle.ViewModel;

import com.fieldservice.app.domain.model.Priority;
import com.fieldservice.app.domain.model.Ticket;
import com.fieldservice.app.domain.model.Technician;
import com.fieldservice.app.domain.repository.AuthRepository;
import com.fieldservice.app.domain.repository.TicketRepository;
import com.fieldservice.app.presentation.tickets.TicketFilter;

import java.util.ArrayList;
import java.util.List;

public class HomeViewModel extends ViewModel {

    private static final int MAX_PRIORITY_TICKETS = 3;

    private final LiveData<HomeUiState> uiState;

    public HomeViewModel(TicketRepository ticketRepository, AuthRepository authRepository) {
        String firstName = firstNameOf(authRepository.getLoggedInTechnician());
        uiState = Transformations.map(
                ticketRepository.observeTickets(),
                tickets -> buildState(tickets, firstName)
        );
    }

    public LiveData<HomeUiState> getUiState() {
        return uiState;
    }

    private static String firstNameOf(Technician technician) {
        if (technician == null || technician.getName() == null) {
            return "";
        }
        String name = technician.getName();
        int spaceIndex = name.indexOf(' ');
        return spaceIndex >= 0 ? name.substring(0, spaceIndex) : name;
    }

    private static HomeUiState buildState(List<Ticket> tickets, String firstName) {
        int pending = 0;
        int inProgress = 0;
        int completed = 0;
        List<Ticket> priorityTickets = new ArrayList<>();

        for (Ticket ticket : tickets) {
            if (TicketFilter.PENDING.matches(ticket.getStatus())) {
                pending++;
            }
            if (TicketFilter.IN_PROGRESS.matches(ticket.getStatus())) {
                inProgress++;
            }
            if (TicketFilter.COMPLETED.matches(ticket.getStatus())) {
                completed++;
            }

            boolean notFinished = TicketFilter.PENDING.matches(ticket.getStatus())
                    || TicketFilter.IN_PROGRESS.matches(ticket.getStatus());
            boolean highPriority = ticket.getPriority() == Priority.HIGH
                    || ticket.getPriority() == Priority.CRITICAL;
            if (notFinished && highPriority) {
                priorityTickets.add(ticket);
            }
        }

        priorityTickets.sort((a, b) -> b.getPriority().compareTo(a.getPriority()));
        if (priorityTickets.size() > MAX_PRIORITY_TICKETS) {
            priorityTickets = priorityTickets.subList(0, MAX_PRIORITY_TICKETS);
        }

        return new HomeUiState(firstName, pending, inProgress, completed, priorityTickets);
    }
}

File: legacy-kotlin/app/src/main/java/com/fieldservice/app/presentation/home/HomeViewModel.kt
package com.fieldservice.app.presentation.home

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.fieldservice.app.domain.model.Priority
import com.fieldservice.app.domain.model.Ticket
import com.fieldservice.app.domain.repository.AuthRepository
import com.fieldservice.app.domain.repository.TicketRepository
import com.fieldservice.app.presentation.tickets.TicketFilter
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.catch
import kotlinx.coroutines.flow.launchIn
import kotlinx.coroutines.flow.onEach

data class HomeUiState(
    val isLoading: Boolean = true,
    val technicianFirstName: String = "",
    val pendingCount: Int = 0,
    val inProgressCount: Int = 0,
    val completedCount: Int = 0,
    val priorityTickets: List<Ticket> = emptyList(),
    val errorMessage: String? = null
)

private val NOT_FINISHED_STATUSES = TicketFilter.PENDING_STATUSES + TicketFilter.IN_PROGRESS_STATUSES

class HomeViewModel(
    ticketRepository: TicketRepository,
    authRepository: AuthRepository
) : ViewModel() {

    private val _uiState = MutableStateFlow(
        HomeUiState(technicianFirstName = authRepository.getLoggedInTechnician()?.name?.substringBefore(" ").orEmpty())
    )
    val uiState: StateFlow<HomeUiState> = _uiState

    init {
        ticketRepository.observeTickets()
            .onEach { tickets ->
                _uiState.value = _uiState.value.copy(
                    isLoading = false,
                    pendingCount = tickets.count { TicketFilter.PENDING.matches(it.status) },
                    inProgressCount = tickets.count { TicketFilter.IN_PROGRESS.matches(it.status) },
                    completedCount = tickets.count { TicketFilter.COMPLETED.matches(it.status) },
                    priorityTickets = tickets
                        .filter { it.status in NOT_FINISHED_STATUSES }
                        .filter { it.priority == Priority.HIGH || it.priority == Priority.CRITICAL }
                        .sortedByDescending { it.priority }
                        .take(3),
                    errorMessage = null
                )
            }
            .catch { error ->
                _uiState.value = _uiState.value.copy(isLoading = false, errorMessage = error.message)
            }
            .launchIn(viewModelScope)
    }
}
